File: python/sunlight/capture.py
"""Cooperative creation-hook capture (spec §4.2, §5).

State machine:
  NEW → STAGING → FINALIZED → HASHED → SIGNED → PUBLISHED
  NEW/STAGING/FINALIZED/HASHED → FAILED        (local.CaptureFailed)
  SIGNED → FAILED                             (local.PublishFailed)
  SIGNED → PUBLICATION_UNKNOWN → PUBLISHED|FAILED
  FAILED → STAGING only via an explicit new capture() call.
"""

from __future__ import annotations

import os
import secrets
import shutil
import stat as statmod

from .artifact import enumerate_tree, hash_file_bytes, manifest_artifact
from .domains import B
from .errors import SunlightError
from .fsutil import dir_fsync, write_file_exclusive
from .jcs import jcs
from .objects import Signer, sign_statement
from .schema import parse_statement_body, parse_tree_manifest
from .strictjson import parse_strict_json

STAGED_NAME = "staged"


def files_identical(a: str, b: str) -> bool:
    sa, sb = os.lstat(a), os.lstat(b)
    if not (statmod.S_ISREG(sa.st_mode) and statmod.S_ISREG(sb.st_mode)) or sa.st_size != sb.st_size:
        return False
    return hash_file_bytes(a)["digest"] == hash_file_bytes(b)["digest"]


def _read_all(path: str) -> bytes:
    with open(path, "rb") as f:
        return f.read()


def capture(request: dict, produce, signer: Signer, journal=lambda event, detail=None: None):
    staging_parent = request["staging_parent"]
    publish_path = request["publish_path"]
    profile = request["profile"]
    template = request["template"]

    if not os.path.isdir(staging_parent) or os.path.islink(staging_parent):
        raise SunlightError("USAGE", cause=ValueError("staging_parent is not a directory"))
    if os.path.lexists(publish_path):
        raise SunlightError("OUTPUT_EXISTS")

    staging = os.path.join(staging_parent, f".sunlight-stage-{secrets.token_hex(8)}")

    try:
        os.mkdir(staging, 0o700)
    except OSError as e:
        journal("local.CaptureFailed")
        raise SunlightError("IO_ERROR", cause=e) from e
    journal("local.CaptureStarted")

    staged_path = os.path.join(staging, STAGED_NAME)
    cleanup = lambda: shutil.rmtree(staging, ignore_errors=True)

    # STAGING → FINALIZED
    try:
        if profile == "tree/1":
            os.mkdir(staged_path, 0o700)
        produce(staged_path)
        if profile == "tree/1":
            if not os.path.isdir(staged_path):
                raise SunlightError("IO_ERROR", cause=ValueError("producer did not create a directory"))
        elif not os.path.isfile(staged_path):
            raise SunlightError("IO_ERROR", cause=ValueError("producer did not create a file"))
    except Exception as e:
        journal("local.CaptureFailed")
        cleanup()
        raise e if isinstance(e, SunlightError) else SunlightError("IO_ERROR", cause=e)
    journal("local.BytesFinalized")

    # FINALIZED → HASHED
    publish_bytes = None
    try:
        if profile == "bytes/1":
            r = hash_file_bytes(staged_path)
            artifact = {"profile": profile, "digest": r["digest"], "bytes": r["bytes"]}
        elif profile == "jcs/1":
            if os.lstat(staged_path).st_size > 1024 * 1024:
                raise SunlightError("BODY_LIMIT")
            try:
                value = parse_strict_json(_read_all(staged_path))
            except Exception as e:
                raise SunlightError("SCHEMA_INVALID", cause=e) from e
            publish_bytes = jcs(value)
            artifact = {"profile": profile, "digest": B(publish_bytes), "bytes": len(publish_bytes)}
        else:
            rels = enumerate_tree(staged_path)
            files = [
                {"path": rel, **hash_file_bytes(os.path.join(staged_path, *rel.split("/")))}
                for rel in rels
            ]
            manifest = parse_tree_manifest({"v": "sunlight.tree/1", "files": files})
            artifact = manifest_artifact(manifest)
    except Exception as e:
        journal("local.CaptureFailed")
        cleanup()
        raise e if isinstance(e, SunlightError) else SunlightError("IO_ERROR", cause=e)
    journal("local.ArtifactHashed", {"digest": artifact["digest"]})

    # HASHED → SIGNED
    try:
        body = parse_statement_body({
            "v": template["v"],
            "id": template["id"],
            "ledger": template["ledger"],
            "signer": template["signer"],
            "claimed_at_ms": template["claimed_at_ms"],
            "capture": "creation_hook",
            "subject": {"kind": template["kind"], "artifact": artifact},
            "parents": template["parents"],
            "details": template["details"],
            "evidence": template["evidence"],
        })
        statement = sign_statement(body, signer)
        write_file_exclusive(os.path.join(staging, "statement.json"), jcs(statement))
    except Exception as e:
        journal("local.CaptureFailed")
        cleanup()
        raise e if isinstance(e, SunlightError) else SunlightError("IO_ERROR", cause=e)
    journal("local.StatementSigned", {"statement": statement["hash"]})

    # SIGNED → PUBLISHED
    try:
        if publish_bytes is not None:
            write_file_exclusive(publish_path, publish_bytes)
        else:
            os.rename(staged_path, publish_path)
        dir_fsync(os.path.dirname(publish_path) or ".")
    except Exception as e:
        if not os.path.lexists(publish_path):
            journal("local.PublishFailed")
            raise SunlightError("IO_ERROR", cause=e) from e
        journal("local.PublicationUncertain")
        try:
            if profile != "tree/1":
                expected = publish_bytes if publish_bytes is not None else _read_all(staged_path)
                if expected != _read_all(publish_path):
                    raise SunlightError("ARTIFACT_CHANGED")
            elif not os.path.isdir(publish_path):
                raise SunlightError("ARTIFACT_CHANGED")
            os.lstat(os.path.join(staging, "statement.json"))
            journal("local.PublicationRecovered")
            return {"published_path": publish_path, "statement": statement, "recorded": False}
        except Exception as e2:
            journal("local.PublishFailed")
            raise e2 if isinstance(e2, SunlightError) else SunlightError("IO_ERROR", cause=e2)

    journal("local.ArtifactPublished", {"path": publish_path})
    return {"published_path": publish_path, "statement": statement, "recorded": False}

File: src/capture.ts
/**
 * Cooperative creation-hook capture (spec §4.2, §5).
 *
 * State machine:
 *   NEW → STAGING → FINALIZED → HASHED → SIGNED → PUBLISHED
 *   NEW/STAGING/FINALIZED/HASHED → FAILED        (local.CaptureFailed)
 *   SIGNED → FAILED                             (local.PublishFailed)
 *   SIGNED → PUBLICATION_UNKNOWN → PUBLISHED|FAILED
 *   FAILED → STAGING only via an explicit new capture() call.
 *
 * The wrapper returns the immutable published path and the signed statement
 * together; sign precedes exposing the artifact through the publication path.
 * `recorded` is always false here: registry acknowledgement is a separate
 * state, never a fabricated timestamp.
 */

import {
  mkdirSync, lstatSync, renameSync, openSync, closeSync, writeSync,
  fsyncSync, readSync, statSync, rmSync,
} from "node:fs";
import { join } from "node:path";
import { randomBytes } from "node:crypto";
import { SunlightError } from "./errors.js";
import { jcsBytes } from "./encoding/jcs.js";
import { parseStrictJson } from "./encoding/strict-json.js";
import { B } from "./crypto/domains.js";
import { enumerateTree, hashFileBytes, manifestArtifact } from "./artifact.js";
import {
  parseStatementBody, parseTreeManifest,
  type Artifact, type ArtifactProfile, type Statement, type StatementBody,
} from "./schema.js";
import { signStatement, type Signer } from "./objects.js";

export type CaptureState =
  | "NEW" | "STAGING" | "FINALIZED" | "HASHED" | "SIGNED"
  | "PUBLISHED" | "FAILED" | "PUBLICATION_UNKNOWN";

export type Journal = (event: string, detail?: Record<string, unknown>) => void;

export interface CaptureTemplate {
  v: "sunlight.statement/1";
  id: string;
  ledger: string;
  signer: string;
  claimed_at_ms: number;
  kind: "dataset" | "model" | "evidence";
  parents: StatementBody["parents"];
  details: StatementBody["details"];
  evidence: StatementBody["evidence"];
}

export interface CaptureRequest {
  staging_parent: string;
  publish_path: string;
  profile: ArtifactProfile;
  template: CaptureTemplate;
}

export interface CaptureResult {
  published_path: string;
  statement: Statement;
  recorded: false;
}

const STAGED_NAME = "staged";

function writeFileSyncExclusive(path: string, data: Buffer): void {
  const fd = openSync(path, "wx", 0o600);
  try {
    writeSync(fd, data);
    fsyncSync(fd);
  } finally {
    closeSync(fd);
  }
}

function readAll(path: string): Buffer {
  const st = statSync(path);
  const fd = openSync(path, "r");
  try {
    const b = Buffer.allocUnsafe(st.size);
    let off = 0;
    while (off < st.size) {
      const n = readSync(fd, b, off, st.size - off, null);
      if (n === 0) break;
      off += n;
    }
    return b.subarray(0, off);
  } finally {
    closeSync(fd);
  }
}

function dirFsync(path: string): void {
  const fd = openSync(path, "r");
  try {
    fsyncSync(fd);
  } finally {
    closeSync(fd);
  }
}

/** Byte-compare two files without loading either fully. */
export function filesIdentical(a: string, b: string): boolean {
  const sa = lstatSync(a);
  const sb = lstatSync(b);
  if (!sa.isFile() || !sb.isFile() || sa.size !== sb.size) return false;
  return hashFileBytes(a).digest === hashFileBytes(b).digest;
}

export async function capture(
  request: CaptureRequest,
  produce: (stagedPath: string) => Promise<void>,
  signer: Signer,
  journal: Journal = () => {},
): Promise<CaptureResult> {
  const { staging_parent, publish_path, profile, template } = request;
  let state: CaptureState = "NEW";

  const parentStat = lstatSync(staging_parent);
  if (!parentStat.isDirectory()) {
    throw new SunlightError("USAGE", null, new Error("staging_parent is not a directory"));
  }
  if (lstatSync(publish_path, { throwIfNoEntry: false }) !== undefined) {
    throw new SunlightError("OUTPUT_EXISTS");
  }

  const staging = join(
    staging_parent,
    `.sunlight-stage-${randomBytes(8).toString("hex")}`,
  );

  // NEW → STAGING
  try {
    mkdirSync(staging, { mode: 0o700 });
  } catch (e) {
    journal("local.CaptureFailed");
    throw new SunlightError("IO_ERROR", null, e);
  }
  state = "STAGING";
  journal("local.CaptureStarted");

  const stagedPath = join(staging, STAGED_NAME);
  const cleanup = (): void => {
    try {
      rmSync(staging, { recursive: true, force: true });
    } catch { /* best-effort cleanup */ }
  };

  // STAGING → FINALIZED: the producer fills the staged path exclusively.
  try {
    if (profile === "tree/1") {
      mkdirSync(stagedPath, { mode: 0o700 });
    }
    await produce(stagedPath);
    if (profile === "tree/1") {
      if (!lstatSync(stagedPath).isDirectory()) {
        throw new SunlightError("IO_ERROR", null, new Error("producer did not create a directory"));
      }
    } else if (!lstatSync(stagedPath).isFile()) {
      throw new SunlightError("IO_ERROR", null, new Error("producer did not create a file"));
    }
  } catch (e) {
    state = "FAILED";
    journal("local.CaptureFailed");
    cleanup();
    throw e instanceof SunlightError ? e : new SunlightError("IO_ERROR", null, e);
  }
  state = "FINALIZED";
  journal("local.BytesFinalized");

  // FINALIZED → HASHED
  let artifact: Artifact;
  let publishBytes: Buffer | null = null; // jcs/1 publishes canonical bytes
  try {
    if (profile === "bytes/1") {
      const r = hashFileBytes(stagedPath);
      artifact = { profile, digest: r.digest, bytes: r.bytes };
    } else if (profile === "jcs/1") {
      const st = lstatSync(stagedPath);
      if (st.size > 1024 * 1024) throw new SunlightError("BODY_LIMIT");
      let value: unknown;
      try {
        value = parseStrictJson(readAll(stagedPath));
      } catch (e) {
        throw new SunlightError("SCHEMA_INVALID", null, e);
      }
      publishBytes = jcsBytes(value);
      artifact = { profile, digest: B(publishBytes), bytes: publishBytes.length };
    } else {
      // tree/1 — the producer populated the staged directory directly; the
      // staged files are already the private copies, so manifest them in place.
      const rels = enumerateTree(stagedPath);
      const files = rels.map((rel) => {
        const { digest, bytes } = hashFileBytes(join(stagedPath, ...rel.split("/")));
        return { path: rel, digest, bytes };
      });
      const manifest = parseTreeManifest({ v: "sunlight.tree/1", files });
      artifact = manifestArtifact(manifest);
    }
  } catch (e) {
    state = "FAILED";
    journal("local.CaptureFailed");
    cleanup();
    throw e instanceof SunlightError ? e : new SunlightError("IO_ERROR", null, e);
  }
  state = "HASHED";
  journal("local.ArtifactHashed", { digest: artifact.digest });

  // HASHED → SIGNED
  let statement: Statement;
  try {
    const body = parseStatementBody({
      v: template.v,
      id: template.id,
      ledger: template.ledger,
      signer: template.signer,
      claimed_at_ms: template.claimed_at_ms,
      capture: "creation_hook",
      subject: { kind: template.kind, artifact },
      parents: template.parents,
      details: template.details,
      evidence: template.evidence,
    });
    statement = await signStatement(body, signer);
    // Persist the signed statement durably inside staging before publication.
    writeFileSyncExclusive(join(staging, "statement.json"), jcsBytes(statement));
  } catch (e) {
    state = "FAILED";
    journal("local.CaptureFailed");
    cleanup();
    throw e instanceof SunlightError ? e : new SunlightError("IO_ERROR", null, e);
  }
  state = "SIGNED";
  journal("local.StatementSigned", { statement: statement.hash });

  // SIGNED → PUBLISHED: atomic rename of the staged artifact.
  try {
    if (publishBytes !== null) {
      writeFileSyncExclusive(publish_path, publishBytes);
    } else {
      renameSync(stagedPath, publish_path);
    }
    dirFsync(join(publish_path, ".."));
  } catch (e) {
    // Publication may or may not have completed.
    const published = lstatSync(publish_path, { throwIfNoEntry: false });
    if (published === undefined) {
      state = "FAILED";
      journal("local.PublishFailed");
      // The signed statement is retained in staging as the recovery object.
      throw new SunlightError("IO_ERROR", null, e);
    }
    state = "PUBLICATION_UNKNOWN";
    journal("local.PublicationUncertain");
    // Recovery: reopen the destination and verify it carries the exact
    // staged bytes and that the statement is durable.
    try {
      if (profile !== "tree/1") {
        const expected = publishBytes ?? readAll(stagedPath);
        const got = readAll(publish_path);
        if (!expected.equals(got)) throw new SunlightError("ARTIFACT_CHANGED");
      } else if (!lstatSync(publish_path).isDirectory()) {
        throw new SunlightError("ARTIFACT_CHANGED");
      }
      lstatSync(join(staging, "statement.json"));
      state = "PUBLISHED";
      journal("local.PublicationRecovered");
      return { published_path: publish_path, statement, recorded: false };
    } catch (e2) {
      state = "FAILED";
      journal("local.PublishFailed");
      throw e2 instanceof SunlightError ? e2 : new SunlightError("IO_ERROR", null, e2);
    }
  }
  state = "PUBLISHED";
  journal("local.ArtifactPublished", { path: publish_path });
  return { published_path: publish_path, statement, recorded: false };
}
